src/wasm.rs

```diff
@@ -0,0 +1,129 @@
+use std::{
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+use futures::{stream::SplitSink, Sink, Stream, StreamExt};
+use pharos::{Observable, ObserveConfig};
+use pin_project_lite::pin_project;
+use ws_stream_wasm::{WsErr, WsEvent, WsMessage, WsMeta, WsStream};
+
+use crate::websockets::WebsocketMessage;
+
+/// Creates a new pair of stream and sink which operate on [`WasmWebsocketMessage`] instead of `WsMessage` and `WsEvent` separately.
+pub async fn wasm_websocket_combined_split(
+    mut ws_meta: WsMeta,
+    ws_stream: WsStream,
+) -> (
+    impl Stream<Item = Result<WasmWebsocketMessage, WsErr>>,
+    impl Sink<WasmWebsocketMessage, Error = WsErr>,
+) {
+    let event_stream = ws_meta.observe(ObserveConfig::default()).await.unwrap();
+
+    let (sink, message_stream) = ws_stream.split();
+
+    let message_stream = message_stream.map(|f| Ok::<_, WsErr>(WasmWebsocketMessage::WsMessage(f)));
+
+    let event_stream = event_stream.map(|f| match f {
+        WsEvent::WsErr(e) => Err(e),
+        e => Ok(WasmWebsocketMessage::WsEvent(e)),
+    });
+
+    let stream = futures::stream::select(event_stream, message_stream);
+
+    let sink = FusedWasmWebsocketSink::new(sink);
+
+    (stream, sink)
+}
+
+#[derive(Debug)]
+/// A WebSocket event abstraction to combine the differentiated WebSocket events and messages of `ws_stream_wasm`.
+pub enum WasmWebsocketMessage {
+    /// A WebSocket message, text or binary, was received.
+    WsMessage(ws_stream_wasm::WsMessage),
+    /// A WebSocket event other than a message was received.
+    WsEvent(ws_stream_wasm::WsEvent),
+}
+
+impl WebsocketMessage for WasmWebsocketMessage {
+    type Error = ws_stream_wasm::WsErr;
+
+    fn new(text: String) -> Self {
+        WasmWebsocketMessage::WsMessage(ws_stream_wasm::WsMessage::Text(text))
+    }
+
+    fn text(&self) -> Option<&str> {
+        match self {
+            WasmWebsocketMessage::WsMessage(ws_stream_wasm::WsMessage::Text(text)) => {
+                Some(text.as_ref())
+            }
+            _ => None,
+        }
+    }
+
+    fn error_message(&self) -> Option<String> {
+        match self {
+            WasmWebsocketMessage::WsEvent(ws_stream_wasm::WsEvent::WsErr(error)) => {
+                Some(error.to_string())
```

Comment:
Ah, is this the allocation you were talking about on discord?  I thought you were talking about the `text` function. 

Wonder if it would be worth returning an `Option<Cow<'a, str>>` so that we don't force the libs that don't need to allocate to do so.

Reply:
I guess we can do that :)